feat(gomod): add gomodTidyAll option for indirect dependency updates in Go monorepos

ducthinh993 opened the pull request:
## Changes

- Refactored Go module dependency graph logic to use a dotnet/nuget-inspired pattern.
- Added `package-tree.ts` for graph-based dependency resolution.
- Updated `artifacts.ts` to use new dependency graph and topological update order.
- Improved logging and comments to match project conventions.
- Added/updated unit tests for all new logic.

## Context

Closes #12999.  
This PR implements the new `goModTidyAll` logic for Go monorepos, ensuring all impacted modules are updated in dependency order when local replace directives are present.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Newly added/modified unit tests, or
- [ ] Code inspection only, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository https://github.com/ducthinh993/renovate-gomod-indirect-sample/pull/1

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=36848) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=36848) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=36848) it.</sub>

rarkins commented on lib/modules/manager/gomod/artifacts.ts:
```diff
@@ -426,6 +428,42 @@ export async function updateArtifacts({
+
+        if (modifiedGoModFiles.length > 0) {
+          // Only call processImpactChain if goModTidyAll is enabled
+          if (config.postUpdateOptions?.includes('goModTidyAll')) {
```
Move this up to the first if you added

rarkins reviewed (changes requested):
This is adding massive complexity and maintenance burden so I'm not sure about this. 

I assume it was AI-coded?

viceice reviewed (changes requested):
checkout the dotnet manager. we've a graph builder module which is used there to build the dependency tree.

ducthinh993 committed:
feat(gomod): implement gomodTidyAll option to update dependent go.mod files

- Added support for the gomodTidyAll option, which runs `go mod tidy` on all dependent go.mod files with local replace directives after updates.
- Introduced new functions to handle the retrieval and updating of dependent go.mod files.
- Added tests to ensure correct functionality of the new feature and handle various scenarios, including circular references and complex dependency chains.
- Updated documentation to include gomodTidyAll in the list of postUpdateOptions.

ducthinh993 committed:
refactor(gomod): simplify gomod artifact updates and enhance dependency handling

- Removed redundant tests and code related to handling multiple go.mod files and their dependencies.
- Streamlined the logic for updating artifacts by focusing on direct dependencies and eliminating unnecessary complexity.
- Improved handling of replace directives, including support for multiline blocks and comments.
- Updated tests to reflect changes in dependency relationships and ensure robustness against various scenarios.

ducthinh993 committed:
chore(gomod): remove gomod-dependency-tree-analysis and test coverage files

ducthinh993 committed:
refactor(gomod): enhance artifact update logic for go.mod and go.sum

- Simplified the logging messages for updated files to be more concise.
- Streamlined the addition of updated artifacts by directly pushing the results without unnecessary checks.
- Improved error handling to return detailed artifact errors when updates fail.

ducthinh993 committed:
refactor(gomod): improve regex handling for replace directives

ducthinh993 commented:
Hi @rarkins / @viceice, can I get your reviews please? Thanks
